### ISSUE TITLE
Implement a work-queue algorithm to process Equihash partial solutions in parallel

### ISSUE THREAD

**Issue by daira** [zcash/zcash#1239]
The structure of the memory-optimised Equihash solver is that part of the algorithm generates candidate partial solutions, and another part reconstructs the full solution (which then gets difficulty-tested).

For the current n=200, k=9 parameters, generating partial solutions and reconstructing each solution take approximately the same time. (For the previous n=96, k=3 parameters, generating partials took longer than reconstructing each of them.)

A fairly simple and effective approach to reducing solution latency, therefore, is to parallelize the reconstruction across all of the available candidates (@str4d also mentioned this on a forum thread). I suggest the following approach:
- split the existing solver explicitly into two pure-functional components -- one to generate partial solutions and one to reconstruct full solutions. (These can then be tested independently.)
- maintain a shared work queue (a multi-reader multi-writer FIFO queue, such as [boost::lockfree::queue](http://www.boost.org/doc/libs/1_57_0/doc/html/boost/lockfree/queue.html)) of partial solutions. When a thread is idle, it attempts to pop an item from the queue. If the queue was non-empty, it reconstructs the popped partial. If the queue was empty, the thread generates a new set of partials and then pushes them onto the end of the queue.
- when a new block is seen, cancel all threads and empty the work queue.

Note that this always prioritizes assigning idle threads to reconstructing existing partial solutions when available, rather than generating new partial solutions. I think that is the best strategy --given no parallelization within the generation or reconstruction algorithms-- because:
- it minimises the latency to finding the next full solution;
- it should be no worse in throughput. (There's no "wasted" work relative to other strategies.)

Even if we later change the parameters or implementation so that reconstruction takes less time relative to generation, this strategy should still pay off.

**Comment by daira** [comment]
Note that this interacts with benchmarking: I suggest benchmarking the latency of generation, the latency of reconstruction, and the resulting throughput of full solutions separately. That should give a less noisy indication of any improvements or regressions to either latency or throughput.

**Comment by daira** [comment]
The assertion that there is no wasted work relative to other strategies assumes that the probability of any given partial being correct does not vary across generation runs. Do you think that assumption is correct, @str4d?

**Comment by str4d** [comment]
Do you mean a partial being correct as in just a correct Equihash solution, or one that satisfies the difficulty check? The latter is definitely true because of the additional hashing step. The former AFAICT depends on the implementation strategy:

- Prior to adding `IsProbablyDuplicate()`, the potential partials had much more variation across solver runs (tens to hundreds) than the actual expected number of solutions.
- After adding `IsProbablyDuplicate()`, almost every partial is a valid Equihash solution (at least based on what I've seen).

**Comment by daira** [comment]
I meant just a valid Equihash solution.

**Comment by str4d** [comment]
Then yes, with the current parameters, the time to recreate all partials is on average twice the time to find them, and the above implementation strategy should be optimal.

**Comment by daira** [comment]
One way in which it could still not quite be optimal is that it doesn't allow a thread to switch from doing a generation run to reconstructing a partial that has just been found by another thread. But I think this is unlikely to be a serious issue in practice.

**Comment by daira** [comment]
A minor optimization is for the thread that completes a generation run to immediately go on to reconstructing the first solution, and only push the second and subsequent partials on the work queue. This has almost the same effect other than changing the order in which partials are processed and making slightly less use of the queue.

**Comment by nathan-at-least** [comment]
We're going to de-prioritize this for the time being. If, after launch, we don't see a healthy ecosystem of open source mining, then we can reprioritize improvements like this one.

**Comment by daira** [comment]
This is no longer relevant, since Equihash implementations have been optimised to the point that solution latency is no longer an issue. (The default implementation is still very slow, but we'll remove that.)
